src/interrupts/primask_mutex.rs

```diff
@@ -0,0 +1,69 @@
+//! Mutex for interrupt synchronization.
+//!
+//! The mutex uses the `primask` core register of the cortex m processor
+//! to disable interrupts and synchronize the access to shared variables.
+//!
+//! Since the access to the data is synchronized (no interrupt can preempt
+//! the current code in the critical section) the mutex implements `Send` and `Sync` when
+//! the synchronized data implements `Send`.
+
+
+use core::cell::UnsafeCell;
+
+/// Mutex that uses the `primask` core register from the cortem m processor to disable 
+/// interrupts before the critical section and enables interrupts again, when interrupts
+/// were enabled before entering the critical section.
+///
+/// Since the access to the data is synchronized (no interrupt can preempt
+/// the current code in the critical section) the mutex implements `Send` and `Sync` when
+/// the synchronized data implements `Send`.
+pub struct PrimaskMutex<T> {
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Send for PrimaskMutex<T> {}
+unsafe impl<T: Send> Sync for PrimaskMutex<T> {}
+
+impl<T> PrimaskMutex<T> {
+    /// Takes ownership over the data and return a new mutex
+    ///
+    /// # Examples
+    /// ```
+    /// let x = 5;
+    /// let mutex = PrimaskMutex::new(x);
+    /// ```
+    pub fn new(data: T) -> PrimaskMutex<T> {
+        PrimaskMutex { data: UnsafeCell::new(data) }
+    }
+
+    /// Executes the closure `critical_section` without interrupts.
+    /// 
+    /// If interrupts were enabled before entering the critical section, the interrupts are enabled again
+    /// after the critical section
+    ///
+    /// # Examples
+    /// ```
+    /// let x = 5;
+    /// let mutex = PrimaskMutex::new(x);
+    /// mutex.lock(|data| {
+    ///     // Interrupt free section  
+    ///     // Safe, because 'atomic' block
+    ///     data += 5;    
+    /// });
+    /// // Interrupts are enabled again, if interrupts was enabled before the critical section
+    /// ```
+    pub fn lock<F>(&self, critical_section: F) 
+        where F: FnOnce(&mut T)
+    {
+        // PRIMASK = 1 => Prevents the activation of all exceptions with configurable priority
+        let primask = unsafe { ::cortex_m::register::primask::read() } & 1 == 1;
```

Comment:
I don't think we have a data race there.

Imagine we have to activites `A1` and `A2`, where `A2` is an ISR and have a higher priority than `A1`.
1. `A1` reads the primask register and afterwards A1 is preempted.
2. `A2` reads the the primask register, disables interrupts and enters the critical section.
3. `A2` leaves the critical section and enables interrupts again.
3. `A1` enters the critical section.

Since A2 always leaves the critical section (and enable interrupts again) before A1 is conntinued, the Primask register is always in the same state, as it was before A1 was preempted by A2. I think the important thing is, that A2 always returns, before A1 is running again!

Have I overlooked something?

Interrups are disabled by changing the state of the primask register. (Spacial assembler instruction are used to set and reset the primask register, which is one of the core registers).